index/index_monitor.go

```diff
@@ -0,0 +1,46 @@
+package index
+
+// helpers to atomically switch from using one index to another
+
+import (
+	"sync"
+)
+
+type IndexMonitor struct {
```

Comment:
hrm, I think what I don't like is that the functionality is split over three types - it's hard to reason about.

maybe we could compromise and generalize it? There's no reason it has to be specific to indexes - you could pass it an `interface{}` and a `sync.WaitGroup`, maybe. Or have it take a `ReferenceCounted` interface, which has a `AddReference` and `RemoveReference` method. I'm not seeing any great solutions, though. 

can a map have a pointer key? then you could keep the `WaitGroup` in this type.